WalletWasabi/WabiSabi/Client/AliceClient.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class AliceClient
+	{
+		public AliceClient(Guid aliceId, Guid roundId, ArenaClient arenaClient, IEnumerable<Coin> coins, FeeRate feeRate)
+		{
+			AliceId = aliceId;
+			RoundId = roundId;
+			ArenaClient = arenaClient;
+			Coins = coins;
+			FeeRate = feeRate;
+		}
+
+		public Guid AliceId { get; }
+		public Guid RoundId { get; }
+		private ArenaClient ArenaClient { get; }
+		private IEnumerable<Coin> Coins { get; }
+		private FeeRate FeeRate { get; }
+
+		public async Task ConfirmConnectionAsync(TimeSpan confirmInterval, CancellationToken token)
+		{
+			while (!await ConfirmConnectionAsync().ConfigureAwait(false))
+			{
+				await Task.Delay(confirmInterval, token).ConfigureAwait(false);
+			}
+		}
+
+		private async Task<bool> ConfirmConnectionAsync()
+		{
+			var inputWeight = 4 * Constants.P2wpkhInputVirtualSize;
+			var inputRemainingWeights = new[] { (long)ArenaClient.ProtocolMaxWeightPerAlice - inputWeight };
```

Comment:
```suggestion
			var inputRemainingWeights = new[] { (long)ArenaClient.ProtocolMaxWeightPerAlice - Coins.Count() * inputWeight };
```

Reply:
it could make some sense to have a variable like `var coinCount = Coins.Count();` to avoid iterating the enumeration in case we use it somewhere else too.

Reply:
ACK

Reply:
I merged this without re-reviewing. This is missing and it is important.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/5579